File: Website/Src/Models/SaveData/SavedGame.cs
// Copyright (C) Clicker Heroes Tracker. All Rights Reserved.

using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using ICSharpCode.SharpZipLib.Zip.Compression;
using ICSharpCode.SharpZipLib.Zip.Compression.Streams;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace ClickerHeroesTrackerWebsite.Models.SaveData
{
    /// <summary>
    /// Model for top-level save game data.
    /// </summary>
    /// <remarks>
    /// Ensure the logic here stays in sync with savedGame.ts on the client.
    /// </remarks>
    public class SavedGame
    {
        private const string SprinkleAntiCheatCode = "Fe12NAfA3R6z4k0z";

        private const string SprinkleSalt = "af0ik392jrmt0nsfdghy0";

        private const string AndroidPrefix = "ClickerHeroesAccountSO";

        private const int HashLength = 32;

        private const string ZlibHash = "7a990d405d2c6fb93aa8fbb0ec1a3b23";

        private const string DeflateHash = "7e8bb5a89f2842ac4af01b3b7e228592";

        private static readonly Dictionary<string, EncodingAlgorithm> EncodingAlgorithmHashes = new(StringComparer.OrdinalIgnoreCase)
        {
            { ZlibHash, EncodingAlgorithm.Zlib },
            { DeflateHash, EncodingAlgorithm.Deflate },
        };

        private static readonly Dictionary<EncodingAlgorithm, Func<string, TextReader>> DecodeFuncs = new()
        {
            { EncodingAlgorithm.Sprinkle, DecodeSprinkle },
            { EncodingAlgorithm.Android, DecodeAndroid },
            { EncodingAlgorithm.Zlib, DecodeZlib },
            { EncodingAlgorithm.Deflate, DecodeDeflate },
        };

        private static readonly Dictionary<EncodingAlgorithm, Func<string, string>> EncodeFuncs = new()
        {
            { EncodingAlgorithm.Sprinkle, EncodeSprinkle },
            { EncodingAlgorithm.Android, EncodeAndroid },
            { EncodingAlgorithm.Zlib, EncodeZlib },
            { EncodingAlgorithm.Deflate, EncodeDeflate },
        };

        private readonly EncodingAlgorithm _encoding;

        private SavedGame(EncodingAlgorithm encoding, JObject obj)
        {
            _encoding = encoding;
            Object = obj;
        }

        private enum EncodingAlgorithm
        {
            Unknown,
            Sprinkle,
            Android,
            Zlib,
            Deflate,
        }

        public JObject Object { get; private set; }

        public static SavedGame Parse(string encodedSaveData)
        {
            EncodingAlgorithm encoding = DetermineEncodingAlgorithm(encodedSaveData);

            if (!DecodeFuncs.TryGetValue(encoding, out Func<string, TextReader> decodeFunc))
            {
                return null;
            }

            // Decode the save
            using (TextReader reader = decodeFunc(encodedSaveData))
            {
                if (reader == null)
                {
                    return null;
                }

                // Deserialize the save
                try
                {
                    using (JsonTextReader jsonTextReader = new(reader))
                    {
                        return new SavedGame(encoding, JObject.Load(jsonTextReader));
                    }
                }
                catch (Exception)
                {
                    // Just return null for any invalid save
                    return null;
                }
            }
        }

        public static string ScrubIdentity(string encodedSaveData)
        {
            SavedGame savedGame = Parse(encodedSaveData);
            if (savedGame == null)
            {
                return null;
            }

            // Based on https://github.com/Legocro/Clan-stripper/blob/master/script.js
            savedGame.Object.Remove("type");
            SetPropertyIfExists("email", string.Empty);
            SetPropertyIfExists("passwordHash", string.Empty);
            SetPropertyIfExists("prevLoginTimestamp", 0);
            savedGame.Object.Remove("account");
            SetPropertyIfExists("accountId", 0);
            SetPropertyIfExists("loginValidated", false);
            SetPropertyIfExists("uniqueId", string.Empty);

            void SetPropertyIfExists(string propertyName, JToken value)
            {
                JProperty property = savedGame.Object.Property(propertyName, StringComparison.OrdinalIgnoreCase);
                if (property != null)
                {
                    property.Value = value;
                }
            }

            if (!EncodeFuncs.TryGetValue(savedGame._encoding, out Func<string, string> encodeFunc))
            {
                throw new InvalidOperationException($"Could not find an encoding function for {savedGame._encoding}");
            }

            string newJson = savedGame.Object.ToString(Formatting.None);
            return encodeFunc(newJson);
        }

        private static EncodingAlgorithm DetermineEncodingAlgorithm(string content)
        {
            if (content == null || content.Length < HashLength)
            {
                return EncodingAlgorithm.Unknown;
            }

            // Read the first 32 characters as they are the MD5 hash of the used algorithm
            string encodingAlgorithmHash = content.Substring(0, HashLength);

            // Test if the MD5 hash header corresponds to a known encoding algorithm
            if (EncodingAlgorithmHashes.TryGetValue(encodingAlgorithmHash, out EncodingAlgorithm encodingAlgorithm))
            {
                return encodingAlgorithm;
            }

            // Legacy encodings
            return content.Contains(AndroidPrefix, StringComparison.OrdinalIgnoreCase)
                ? EncodingAlgorithm.Android
                : EncodingAlgorithm.Sprinkle;
        }

        private static TextReader DecodeSprinkle(string encodedSaveData)
        {
            int antiCheatCodeIndex = encodedSaveData.IndexOf(SprinkleAntiCheatCode, StringComparison.Ordinal);
            if (antiCheatCodeIndex == -1)
            {
                // Couldn't find anti-cheat
                return null;
            }

            // Remove every other character, AKA "unsprinkle".
            // Handle odd lengthed strings even though it shouldn't happen.
            char[] unsprinkledChars = new char[(antiCheatCodeIndex / 2) + (antiCheatCodeIndex % 2)];
            for (int i = 0; i < antiCheatCodeIndex; i += 2)
            {
                unsprinkledChars[i / 2] = encodedSaveData[i];
            }

            // Validation
            int expectedHashStart = antiCheatCodeIndex + SprinkleAntiCheatCode.Length;
            char[] saltedChars = new char[unsprinkledChars.Length + SprinkleSalt.Length];
            unsprinkledChars.CopyTo(saltedChars, 0);
            SprinkleSalt.CopyTo(0, saltedChars, unsprinkledChars.Length, SprinkleSalt.Length);
            using (MD5 md5 = MD5.Create())
            {
                byte[] data = md5.ComputeHash(Encoding.UTF8.GetBytes(saltedChars));
                for (int i = 0; i < data.Length; i++)
                {
                    int expectedHashPartIndex = expectedHashStart + (i * 2);
                    string actualHashPart = data[i].ToString("x2");
                    if (actualHashPart[0] != encodedSaveData[expectedHashPartIndex]
                        || actualHashPart[1] != encodedSaveData[expectedHashPartIndex + 1])
                    {
                        // Hash didn't match
                        return null;
                    }
                }
            }

            // Decode
            byte[] bytes = Convert.FromBase64CharArray(unsprinkledChars, 0, unsprinkledChars.Length);

            // Wrap in a text reader
            return new StreamReader(new MemoryStream(bytes));
        }

        private static string EncodeSprinkle(string json)
        {
            byte[] bytes = Encoding.UTF8.GetBytes(json);
            string data = Convert.ToBase64String(bytes);

            StringBuilder sb = new((2 * data.Length) + SprinkleAntiCheatCode.Length);

            // Inject an arbitrary character every other character, AKA "sprinkle".
            for (int i = 0; i < data.Length; i++)
            {
                sb.Append(data[i]);
                sb.Append('0');
            }

            sb.Append(SprinkleAntiCheatCode);

            char[] hashChars = new char[data.Length + SprinkleSalt.Length];
            data.CopyTo(0, hashChars, 0, data.Length);
            SprinkleSalt.CopyTo(0, hashChars, data.Length, SprinkleSalt.Length);
            using (MD5 md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(hashChars));
                for (int i = 0; i < hash.Length; i++)
                {
                    sb.Append(hash[i].ToString("x2"));
                }
            }

            return sb.ToString();
        }

        private static TextReader DecodeAndroid(string encodedSaveData)
        {
            // Get the index of the first open brace
            const char BraceCharCode = '{';
            int firstBrace = encodedSaveData.IndexOf(BraceCharCode, StringComparison.Ordinal);
            if (firstBrace < 0)
            {
                return null;
            }

            string json = encodedSaveData
                .Substring(firstBrace)
                .Replace("\\\"", "\"", StringComparison.Ordinal)
                .Replace("\\\\", "\\", StringComparison.Ordinal);

            // Wrap in a text reader
            return new StringReader(json);
        }

        private static string EncodeAndroid(string json)
        {
            string escapedJson = json
                .Replace("\\", "\\\\", StringComparison.Ordinal)
                .Replace("\"", "\\\"", StringComparison.Ordinal);

            // No idea how accurate these chracters are. Just using what was found in one particular save.
            return "?D?TCSO" + AndroidPrefix + "\tjson??%" + escapedJson;
        }

        private static TextReader DecodeZlib(string encodedSaveData)
        {
            string data = encodedSaveData.Substring(HashLength);
            byte[] bytes = Convert.FromBase64String(data);
            MemoryStream memStream = new(bytes);
            InflaterInputStream inflaterStream = new(memStream);
            return new StreamReader(inflaterStream);
        }

        private static string EncodeZlib(string json)
        {
            using (MemoryStream compressedStream = new(16 * 1024))
            {
                using (MemoryStream jsonStream = new(Encoding.UTF8.GetBytes(json)))
                using (DeflaterOutputStream deflaterStream = new(compressedStream))
                {
                    jsonStream.CopyTo(deflaterStream);
                }

                return ZlibHash + Convert.ToBase64String(compressedStream.ToArray());
            }
        }

        private static TextReader DecodeDeflate(string encodedSaveData)
        {
            string data = encodedSaveData.Substring(HashLength);
            byte[] bytes = Convert.FromBase64String(data);
            MemoryStream memStream = new(bytes);
            InflaterInputStream inflaterStream = new(memStream, new Inflater(noHeader: true));
            return new StreamReader(inflaterStream);
        }

        private static string EncodeDeflate(string json)
        {
            using (MemoryStream compressedStream = new(16 * 1024))
            {
                using (MemoryStream jsonStream = new(Encoding.UTF8.GetBytes(json)))
                using (DeflaterOutputStream deflaterStream = new(compressedStream, new Deflater(level: Deflater.DEFAULT_COMPRESSION, noZlibHeaderOrFooter: true)))
                {
                    jsonStream.CopyTo(deflaterStream);
                }

                return DeflateHash + Convert.ToBase64String(compressedStream.ToArray());
            }
        }
    }
}

File: WebClient/src/models/savedGame.ts
import { inflate, deflate } from "pako";
import * as MD5 from "crypto-js/md5";

export interface IItemData {
    bonusType1: number;
    bonus1Level: string;
    bonusType2: number;
    bonus2Level: string;
    bonusType3: number;
    bonus3Level: string;
    bonusType4: number;
    bonus4Level: string;
}

export interface IItemsData {
    slots: { [id: string]: number };
    items: { [id: string]: IItemData };
}

export interface IAncientData {
    id: number;
    level: number | string;
    spentHeroSouls: number | string;
    purchaseTime: number;
}

export interface IAncientsData {
    ancients: { [id: string]: IAncientData };
}

export interface IOutsiderData {
    id: number;
    level: number;
}

export interface IOutsidersData {
    outsiders: { [id: string]: IOutsiderData };
}

export interface ISavedGameData {
    transcendent: boolean;
    paidForRubyMultiplier: boolean;
    saveOrigin: string;
    unixTimestamp: number;

    primalSouls: number | string;
    heroSouls: number | string;
    heroSoulsSacrificed: number | string;
    ancientSoulsTotal: number | string;
    titanDamage: number | string;
    highestFinishedZonePersist: number | string;
    pretranscendentHighestFinishedZone: number | string;
    transcendentHighestFinishedZone: number | string;
    numAscensionsThisTranscension: number | string;
    numWorldResets: number | string;
    rubies: number | string;
    autoclickers: number | string;
    dlcAutoclickers: number | string;

    items: IItemsData;
    ancients: IAncientsData;
    outsiders: IOutsidersData;
}

type EncodingAlgorithm = "unknown" | "sprinkle" | "android" | "zlib";

// Ensure the logic here stays in sync with SavedGame.cs on the server
export class SavedGame {
    private static readonly sprinkleAntiCheatCode = "Fe12NAfA3R6z4k0z";

    private static readonly sprinkleSalt = "af0ik392jrmt0nsfdghy0";

    private static readonly androidPrefix = "ClickerHeroesAccountSO";

    private static readonly hashLength = 32;

    private static readonly zlibHash = "7a990d405d2c6fb93aa8fbb0ec1a3b23";

    private static readonly encodingAlgorithmHashes: { [hash: string]: EncodingAlgorithm } = {
        [SavedGame.zlibHash]: "zlib",
    };

    private static readonly decodeFuncs: { [encodingAlgorithm: string]: (content: string) => ISavedGameData | null } = {
        sprinkle: SavedGame.decodeSprinkle,
        android: SavedGame.decodeAndroid,
        zlib: SavedGame.decodeZlib,
    };

    private static readonly encodeFuncs: { [encodingAlgorithm: string]: (data: ISavedGameData) => string } = {
        sprinkle: SavedGame.encodeSprinkle,
        android: SavedGame.encodeAndroid,
        zlib: SavedGame.encodeZlib,
    };

    public data: ISavedGameData | null;

    private encoding: EncodingAlgorithm;

    private _scrubbedContent: string | null = null;

    constructor(
        public content: string | null,
        public isScrubbed: boolean,
    ) {
        this.encoding = SavedGame.determineEncodingAlgorithm(this.content);
        this.data = SavedGame.decode(this.content, this.encoding);

        if (this.isScrubbed) {
            this._scrubbedContent = content;
        }
    }

    public get scrubbedContent(): string | null {
        if (!this._scrubbedContent) {
            // Create a copy of the data before altering it
            const data = JSON.parse(JSON.stringify(this.data));

            // Based on https://github.com/Legocro/Clan-stripper/blob/master/script.js
            delete data.type;
            data.email = "";
            data.passwordHash = "";
            data.prevLoginTimestamp = 0;
            data.account = null;
            data.accountId = 0;
            data.loginValidated = false;
            data.uniqueId = "";
            data.subscribeEmail = "";

            this._scrubbedContent = SavedGame.encode(data, this.encoding);
        }

        return this._scrubbedContent;
    }

    public clone(): SavedGame {
        // Passing null content since it's faster to copy the private state than to re-parse
        const savedGame = new SavedGame(null, this.isScrubbed);
        savedGame.content = this.content;
        savedGame.encoding = this.encoding;
        savedGame.data = JSON.parse(JSON.stringify(this.data));
        savedGame._scrubbedContent = this._scrubbedContent;
        return savedGame;
    }

    public updateContent(): void {
        this.content = SavedGame.encode(this.data, this.encoding);
        this._scrubbedContent = null;
    }

    private static determineEncodingAlgorithm(content: string | null): EncodingAlgorithm {
        if (content == null || content.length < SavedGame.hashLength) {
            return "unknown";
        }

        // Read the first 32 characters as they are the MD5 hash of the used algorithm
        const encodingAlgorithmHash = content.substring(0, SavedGame.hashLength);

        // Test if the MD5 hash header corresponds to a known encoding algorithm
        const encodingAlgorithm = SavedGame.encodingAlgorithmHashes[encodingAlgorithmHash];
        if (encodingAlgorithm) {
            return encodingAlgorithm;
        }

        // Legacy encodings
        return content.indexOf(SavedGame.androidPrefix) >= 0
            ? "android"
            : "sprinkle";
    }

    private static decode(content: string | null, encoding: EncodingAlgorithm): ISavedGameData | null {
        const decodeFunc = SavedGame.decodeFuncs[encoding];
        return decodeFunc && content ? decodeFunc(content) : null;
    }

    private static encode(data: ISavedGameData | null, encoding: EncodingAlgorithm): string | null {
        const encodeFunc = SavedGame.encodeFuncs[encoding];
        return encodeFunc && data ? encodeFunc(data) : null;
    }

    private static decodeSprinkle(content: string): ISavedGameData | null {
        const pieces = content.split(SavedGame.sprinkleAntiCheatCode);
        if (pieces.length !== 2) {
            // Couldn't find anti-cheat
            return null;
        }

        const data = pieces[0];
        const hash = pieces[1];

        // Remove every other character, AKA "unsprinkle".
        let unsprinkled = "";
        for (let i = 0; i < data.length; i += 2) {
            unsprinkled += data[i];
        }

        // Validation
        const expectedHash = MD5(unsprinkled + SavedGame.sprinkleSalt).toString();
        // tslint:disable-next-line:possible-timing-attack This isn't a security issue
        if (hash !== expectedHash) {
            return null;
        }

        // Decode
        return JSON.parse(atob(unsprinkled));
    }

    private static encodeSprinkle(data: ISavedGameData): string {
        const json = JSON.stringify(data);
        const base64Data = btoa(json);

        // Inject an arbitrary character every other character, AKA "sprinkle".
        let sprinkled = "";
        for (const ch of base64Data) {
            sprinkled += ch;
            sprinkled += "0";
        }

        const hash = CryptoJS.MD5(base64Data + SavedGame.sprinkleSalt).toString();
        return sprinkled + SavedGame.sprinkleAntiCheatCode + hash;
    }

    private static decodeAndroid(content: string): ISavedGameData | null {
        // Get the index of the first open brace
        const firstBrace = content.indexOf("{");
        if (firstBrace < 0) {
            return null;
        }

        const json = content
            .substring(firstBrace)
            .replace("\\\"", "\"")
            .replace("\\\\", "\\");

        return JSON.parse(json);
    }

    private static encodeAndroid(data: ISavedGameData): string {
        const json = JSON.stringify(data)
            .replace("\\", "\\\\")
            .replace("\"", "\\\"");

        // No idea how accurate these chracters are. Just using what was found in one particular save.
        return "?D?TCSO" + SavedGame.androidPrefix + "\tjson??%" + json;
    }

    private static decodeZlib(content: string): ISavedGameData {
        const result = content.slice(SavedGame.hashLength);
        const data = inflate(atob(result), { to: "string" });
        return JSON.parse(data);

    }

    private static encodeZlib(data: ISavedGameData): string {
        const json = JSON.stringify(data);
        const encodedData = deflate(json, { to: "string" });
        return SavedGame.zlibHash + btoa(encodedData);
    }
}
